### PR TITLE
register transaction

### PR DESCRIPTION
**Pull request by alex-arugon**

### CODE REVIEW

File: core/repository/account.ts [hunk 1]
```diff
@@ -6,11 +6,11 @@ class AccountRepo {
     private memoryAccountsByAddress: { [address: number]: Account } = {};
     private memoryAccountsByPublicKey: { [publicKey: string]: Account } = {};
 
-    public add(account: IAccountFilds): Response<void> {
+    public add(account: IAccountFilds): Response<Account> {
```
Comment (axelsheva) [inline core/repository/account.ts]: please change IAccountFilds to AccountModel like a BlockModel

File: core/service/transaction/register.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import Response from 'shared/model/response';
+import AccountRepo from 'core/repository/account';
+import config from 'shared/util/config';
+import BUFFER from 'core/util/buffer';
+import {ITableObject} from 'core/util/common';
+
+class TransactionRegisterService implements ITransactionService<IAssetRegister> {
+
+    async create(trs: Transaction<IAssetRegister>, data: IAssetRegister ): Promise<IAssetRegister> {
+        trs.recipientAddress = null;
```
Comment (macaqueee) [inline core/service/transaction/register.ts]: `recipientAddress` should be optional in TransactionModel, so you dont have to make it `null`.

File: core/service/transaction/register.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    async verify(trs: Transaction<IAssetRegister>, sender: Account): Promise<Response<void>> {
+        const errors = [];
+
+        if (!trs.asset.referral) {
+            errors.push('Missing referral');
+        }
+
+        if (sender) {
```
Comment (macaqueee) [inline core/service/transaction/register.ts]: you always get account for sender. Previously, in BlockService you have `getOrCreateAccount` method, which will provide you account.

File: core/service/transaction/register.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+    calculateFee(trs: Transaction<IAssetRegister>, sender: Account): number {
+        return 0;
+    }
+
+    calculateUndoUnconfirmed(trs: Transaction<IAssetRegister>, sender: Account): void {
+    }
+
+    async applyUnconfirmed(trs: Transaction<IAssetRegister>, sender: Account): Promise<Response<void>> {
+        const sponsor: Account = AccountRepo.getByAddress(trs.asset.referral);
```
Comment (macaqueee) [inline core/service/transaction/register.ts]: rename to `referral`

File: core/service/transaction/register.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        const sponsorsReferrals: Array<Account> =
```
Comment (macaqueee) [inline core/service/transaction/register.ts]: rename to `referrals`

File: core/service/transaction/register.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        const updateResponse: Response<void> =
+            AccountRepo.updateReferralByAddress(targetAccount.address, [sponsor, ...sponsorsReferrals]);
+        if (!updateResponse.success) {
+            return new Response<void>({ errors: [...addAccountResponse.errors, 'Can\'t update account referrals'] });
+        }
+        return new Response<void>();
+    }
+
+    async undoUnconfirmed(trs: Transaction<IAssetRegister>, sender: Account): Promise<Response<void>> {
+        const targetAccount: Account = AccountRepo.getByAddress(trs.asset.referral);
```
Comment (macaqueee) [inline core/service/transaction/register.ts]: you are removing `referralAccount` here.  You need to remove `sender` account.

**Pull request merged**